src/renderers/gl/renderer.cpp

```diff
@@ -642,6 +859,44 @@ void mrg::Renderer::set_output_transform(glm::mat2 const& t)
     }
 }
 
+// Shader that converts colors to grayscale.
+const GLchar* grayscale_src =
+    "uniform sampler2D tex;\n"
+    "vec4 sample_to_rgba(in vec2 texcoord) {\n"
+    "   vec4 col = texture2D(tex, texcoord);\n"
+    //"   float s = (col[0] + col[1] + col[2]) / 3.0;\n"
+    //"   return vec4(s, s, s, col[3]);\n"
+    //"   return mix(vec4(0.5, 0.5, 0.5, 1.0), col, 1.5);\n"
+    "   return col;\n"
+    "}\n";
+
+// Shader that inverts colors.
+const GLchar* invert_src =
+    "uniform sampler2D tex;\n"
+    "vec4 sample_to_rgba(in vec2 texcoord) {\n"
+    "   vec4 col = texture2D(tex, texcoord);\n"
+    "   return vec4(1.0 - col[0], 1.0 - col[1], 1.0 - col[2], col[3]);\n"
+    "}\n";
+
+void mrg::Renderer::set_output_filter(MirOutputFilter filter)
+{
+    GLchar const * filter_src;
+    switch (filter)
+    {
+    default:
+    case mir_output_filter_none:
+        output_filter_shader = nullptr;
+        return;
+    case mir_output_filter_grayscale:
+        filter_src = grayscale_src;
+        break;
+    case mir_output_filter_invert:
+        filter_src = invert_src;
+        break;
+    }
+    output_filter_shader = std::make_unique<OutputFilterShader>(output_surface->size().width.as_value(), output_surface->size().height.as_value(), filter_src);
+}
```

Comment:
Should our API be that prescriptive? Shouldn't we, instead, accept custom shaders set by shells?

That, or have `DefaultBufferCompositorFactory` take the shader?

Reply:
That was what I was initially going for, but I had trouble finding the correct place to have the shader live. So the PR is a bit of a straw man in that I passed the enum all the way to the renderer.

There's also the challenge of how this fits into software rendering, for the common a11y features we want these to work across both GL and software without any special work, but it is also desirable for shells to define something custom. The question is, do these shells make use of this renderer class or do they override all this with their own class?

Reply:
ACK, I wasn't thinking about software rendering in particular.

> I can see the attraction of @Saviq's suggestion of allowing shells to define arbitrary shaders ("night mode" anyone?), but that opens a "can of worms" over what is supported.

Indeed.